src/Microsoft.Framework.Configuration/ConfigurationSection.cs

```diff
@@ -4,113 +4,124 @@
+            : base(sources)
         {
-            _sources = sources;
+            _key = string.Empty;
+            Value = null;
         }
-
-        public string this[string key]
+        public ConfigurationSection([NotNull] IList<IConfigurationSource> sources, [NotNull] string key)
+            : base (sources)
         {
-            get
-            {
-                return Get(key);
-            }
-            set
+            if (string.IsNullOrEmpty(key))
             {
-                Set(key, value);
+                throw new InvalidOperationException(Resources.Error_EmptyKey);
             }
+
+            _key = key;
+            Value = base[_key];
         }
 
-        public IEnumerable<IConfigurationSource> Sources
+        public string Key
         {
             get
             {
-                return _sources;
+                return _key;
             }
         }
 
-        public string Get([NotNull] string key)
+        public string Value
         {
-            string value;
-            return TryGet(key, out value) ? value : null;
+            get; set;
         }
 
-        public bool TryGet([NotNull] string key, out string value)
+        public override string this[[NotNull] string key]
         {
-            // If a key in the newly added configuration source is identical to a key in a 
-            // formerly added configuration source, the new one overrides the former one.
-            // So we search in reverse order, starting with latest configuration source.
-            foreach (var src in _sources.Reverse())
+            get
             {
-                if (src.TryGet(key, out value))
+                var prefix = _key;
+
+                if (!string.IsNullOrEmpty(prefix))
                 {
-                    return true;
+                    prefix = prefix + Constants.KeyDelimiter;
+                }
+                else
+                {
+                    prefix = string.Empty;
                 }
-            }
-            value = null;
-            return false;
-        }
 
-        public void Set([NotNull] string key, [NotNull] string value)
-        {
-            if (!_sources.Any())
-            {
-                throw new InvalidOperationException(Resources.Error_NoSources);
+                return base[prefix + key];
             }
-            foreach (var src in _sources)
+            set
             {
-                src.Set(key, value);
+                if (!Sources.Any())
+                {
+                    throw new InvalidOperationException(Resources.Error_NoSources);
+                }
+
+                var prefix = _key;
+
+                if (!string.IsNullOrEmpty(prefix))
+                {
+                    prefix = prefix + Constants.KeyDelimiter;
+                }
+                else
+                {
+                    prefix = string.Empty;
+                }
+
+                _key = prefix + key;
```

Comment:
I don't think this should really mutante the `_key` of the current node. Typo? Furthermore `_key` should be read-only and only set at construction time.